json-format.md

```diff
@@ -129,34 +129,82 @@ OPTIONAL not omitted attributes MAY be represeted as a `null` JSON value.
 
 ### 3.1. Handling of "data"
 
+The JSON representation of the event "data" payload is determined by the runtime
+type of the `data` content and the value of the [`datacontenttype`
+attribute][datacontenttype].
+
+#### 3.1.1. Payload Serialization
+
 Before taking action, a JSON serializer MUST first determine the runtime data
 type of the `data` content.
 
 If the implementation determines that the type of data is `Binary`, the value
 MUST be represented as a [JSON string][json-string] expression containing the
 [Base64][base64] encoded binary value, and use the member name `data_base64` to
-store it inside the JSON object.
-
-For any other type, the implementation MUST translate the data value into a
-[JSON value][json-value], and use the member name `data` to store it inside the
-JSON object.
-
-Out of this follows that use of the `data` and `data_base64` members is mutually
-exclusive in a JSON serialized CloudEvent.
+store it inside the JSON representation. The `datacontenttype` SHOULD reflect
+the format of the original binary data.
+
+If the type of data is NOT `Binary`, the implementation MUST next determine
+whether the value of the `datacontenttype` attribute is _JSON-oriented_. A
+JSON-oriented content type is one whose subtype is `json` or has a `+json`
+format extension—that is, if it has the form `*/json` or `*/*+json`.
+
+If the `datacontenttype` is JSON-oriented, a JSON serializer MUST translate the
+data value to a [JSON value][json-value], and use the member name `data` to
+store it inside the JSON representation. Note: this translation SHOULD NOT
+perform a JSON serialization step on the data value. For example, data whose
+runtime type corresponds to a [JSON Object][json-object] (such as a "map" or
+"object" type in the runtime) SHOULD be stored directly as a JSON Object, rather
+than as a JSON-formatted string. An implementation MAY fail to serialize the
+event if it is unable to translate the runtime value to a JSON value.
+
+Otherwise, if the `datacontenttype` is NOT JSON-oriented, a JSON serializer
+MUST use the member name `data` to store a string representation of the data
```

Comment:
I disagree: I think it has to be a string. Either the content type indicates JSON, in which case it's encoded as a JSON value, or the content type indicates something else, in which case by virtue of there being a content type in the first place, it implies there's a conversion between a string (or byte sequence, but we've already handled that case with `data_base64`) and a runtime type.

For example, if the event data is indeed and integer, and for some reason the content type does is not JSON, but some "integer" content type (I don't know what that would be, but let's say "application/integer" for argument's sake), then that content type still implies a particular way of representing that integer _in content_, i.e. as a string.

Reply:
I disagree, but we should get other people's takes on it. Aside from anything else, I'd be happy with the argument that only a string is actually useful.

One possible other case to consider, mind you: formats like YAML or protobuf, which have natural JSON object representations. I think it's at least *reasonable* to say "when using a JSON event format and a protobuf message, the natural data representation is the JSON for the message". I'd like that to be discussed, but for that case I think it would actually be better to put the binary representation in base64. I look forward to hearing the results of the discussion :)

Reply:
Furthermore, if we allow non-string representations in this case, it specifically re-introduces the ambiguity that this PR is trying to solve. If, for example, we allow integer, why not also allow full JSON object? And if we allow full JSON object, then cases like the following become ambiguous again:

```
  "datacontenttype" : "text/html",
  "data" : "<p>hello</p>"
```

```
  "datacontenttype" : "text/html",
  "data" : {
    "p" : "hello"
  }
```

```
  "datacontenttype" : "text/html",
  "data" : "&lt;p&gt;hello&lt;/p&gt;"
```